lib/excon/socket.rb

```diff
@@ -237,18 +269,33 @@ def read_nonblock(max_length)
         @backend_eof = true
       end
 
-      ret = if max_length
+      if max_length
         if @read_buffer.empty?
-          nil # EOF met at beginning
+          # EOF met at beginning
+          @eof = @backend_eof
+          nil
         else
-          @read_buffer.slice!(0, max_length)
+          start = @read_offset
+          readable_bytes = buffer_length
+
+          # Ensure that we can seek backwards when reading until a terminator string.
+          # The read offset must never point past the end of the read buffer.
+          @read_offset += max_length > readable_bytes ? readable_bytes : max_length
+          @read_buffer[start...@read_offset]
         end
       else
         # read until EOFError, so return everything
-        @read_buffer.slice!(0, @read_buffer.length)
+        start = @read_offset
+
+        @read_offset = @read_buffer.length
+        @eof = @backend_eof
+
+        @read_buffer[start..]
       end
-      @eof = @backend_eof && @read_buffer.empty?
-      ret
+    end
+
+    def buffer_length
```

Comment:
For clarity, maybe we should think about naming this something like `buffer_remaining` (presuming I'm understanding it's intention). Right now it otherwise gets a little confusing what the difference between this and `@read_buffer.length` is. Does that make sense?

Reply:
Or perhaps `buffer_available`. That being said, I guess maybe flipping the two words in either case might read better too. ie `remaining_buffer` and `available_buffer` both seem like they might be easier to read. I guess all told, I'd maybe lean toward `remaining_buffer` from those options, but I'm open to other possibilities and discussion.

Reply:
Good call. I think of those I personally like `readable_bytes` best. It says something about our intention/expectation about what we want to do with the bytes, vs `buffered_bytes`, which is accurate, but much less specific I think.

Reply:
I went ahead and changed the name to `readable_bytes`. Let me know what you think.